### DISCUSSION TITLE
How do I fix: "The I/O operation has been aborted because of either a thread error or an application request"

### DISCUSSION THREAD

**Discussion by faustrp** [bonsai-rx/bonsai#2069] [Q&A]
[RudiserialString2.zip](https://github.com/user-attachments/files/17986331/RudiserialString2.zip)
Hello community, 

I have run into a problem with the attached Bonsai workflow in which Bonsai collects data from the Neurophotometrics system and an Arduino through a USB hub. During my initial tests of this version of the workflow, it worked as intended with no errors generated, faithfully recording the data from the Arduino. Since those initial tests, when I've tried to start the workflow, it gives me the same Runtime Error every time: "The I/O operation has been aborted because of either a thread exit or an application request." I do not think this was due to unintended modifications of my workflow because the previous version of the workflow that I did not use for testing now exhibits the same malfunction. I tried restarting the Bonsai computer and disabling and re-enabling the COM port corresponding to the Arduino in Windows Device Manager, neither of which worked. Can anyone help me out with this? I can't start my experiments until this is fixed.

**Comment by bruno-f-cruz** [comment]
Is there an operator that lights up red after the crash? It would be helpful to pinpoint the what code is leading to the exception.

**Reply by faustrp** [reply]
Yes, the SerialReadLine operator lights up red.

**Reply by bruno-f-cruz** [reply]
From your workflow I really cant tell what could be going on. Some tests may help. If you make a new workflow with the `CreateSerialPort` and `SerialReadLine` do you still have this error? If so, it may be a problem with your serial device and I would like to try a new one.

**Reply by faustrp** [reply]
When I made a new workflow (attached) I got a different error: "The given port name does not start with COM/com or does not resolve to a valid serial port. Parametername: portName"
[Wheel_clone.bonsai.zip](https://github.com/user-attachments/files/18017606/Wheel_clone.bonsai.zip)

**Reply by bruno-f-cruz** [reply]
The last workflow has all the operators that the first one has. Can you make one with only the `CreateSerialPort` and `SerialReadLine` as I mentioned above? Thanks!

**Reply by faustrp** [reply]
I made a workflow with only those two operators and I got the initial error: "The I/O operation has been aborted because of either a thread exit or an application request."

**Reply by bruno-f-cruz** [reply]
Can you send the workflow over? Also, have you tried it with a different serial device?

**Reply by faustrp** [reply]
[serialtest.bonsai.zip](https://github.com/user-attachments/files/18050903/serialtest.bonsai.zip)

The workflow is attached. I have not tried it with another Arduino because we don't have a spare Arduino in the lab. I'm also trying to troubleshoot with the Arduino serial monitor but I don't have all the information about this device, which was constructed by a former member of the lab. I'm currently waiting to hear back from him.

**Reply by bruno-f-cruz** [reply]
The workflow looks good. 
> I'm also trying to troubleshoot with the Arduino serial monitor but I don't have all the information about this device, which was constructed by a former member of the lab.

This is also a great idea for troubleshooting. Additionally, unplugging the ardiuno for additional hardware and uploading an empty sketch, could also provide useful information.

**Reply by faustrp** [reply]
I successfully tested with Arduino serial monitor and the Arduino faithfully transmits data to the laptop.

**Reply by faustrp** [reply]
What do you mean by "unplugging the Arduino for additional hardware"? How do I upload a sketch if the Arduino is unplugged?
